### DISCUSSION TITLE
Lifetime Boundaries

### DISCUSSION THREAD

**Discussion by GoldenStack** [TodePond/GulfOfMexico#411] [Q&A] [answered]
Hey,

```java
union a() {
    print(x
}

union b() {
    const const x<-50> = 2!
}

union c() {
    const const x<-50> = 3!
}

a()!
b()!
a()!
c()!
a()!
```
What would this code print?
Can line-based lifetimes exist outside their declared scope? If so, can they override others?

If not, are variables limited to their scope, or are lifetimes transferred outside?
```java
union a(b) {
    print(b)! // Second line
    print(b)! // Third line - would this error?
}

const const x<2> = 2!
a(x)! // First line
print(x)! // Second line
print(x)! // Third line - this definitely errors
```

**Comment by GoldenStack** [comment]
From my other issue [here](https://github.com/TodePond/DreamBerd/issues/403) I would assume that lifetimes just denote which lines in the scope are valid (so the first example would error and the second example would error at the last print statement) but I'm not very sure about this.

**Comment by GoldenStack** [comment]
I've realized that the language would likely be impossible (logically impossible, not just difficult) to implement if my previous assumption were false, so I'm just going to assume the aforementioned assumption is correct.

**Comment by TodePond** [comment, marked as answer]
Variables are scoped to their scope. In production, it would print this:

```
union a() {
    print(x
}

union b() {
    const const x<-50> = 2!
}

union c() {
    const const x<-50> = 3!
}

a()! // "x"
b()! 
a()! // "x"
c()!
a()! // "x"
```

Numbers are passed by value:

```
union a(b) {
    print(b)!
    print(b)!
}

const const x<2> = 2!
a(x)! // 2, 2
print(x)! // 2
print(x)! // "x"
```

**Reply by GoldenStack** [reply]
Makes sense, thanks!
(oops, forgot about unquoted strings. my bad)
